### PR TITLE
Search test

### PR DESCRIPTION
**Pull request by thesuzerain**

### CODE REVIEW

File: src/routes/v2_reroute.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            .map_err(|_| failure_http_response())?;
+        let json_value: Value =
+            serde_json::from_slice(&bytes).map_err(|_| failure_http_response())?;
+        Ok(json_value)
+    } else {
+        Err(response)
+    }
+}
+
+pub async fn alter_actix_multipart(
```
Comment (OmegaJak) [inline src/routes/v2_reroute.rs:70]: Worth considering, not sure if it's worth doing - the methods in this file seem like great candidates for more detailed unit-level tests just around them, no integration. Seems like there are probably a lot of edge cases to be tested here, which can be done faster and more precisely than at the integration level

File: migrations/20231005230721_dynamic-fields.sql [hunk 2]
```diff
@@ -0,0 +1,109 @@
+CREATE TABLE games (
+	id int PRIMARY KEY, -- Only used in db
+	name varchar(64),
+  CONSTRAINT unique_game_name UNIQUE (name)
+);
+INSERT INTO games(id, name) VALUES (1, 'minecraft-java');
+INSERT INTO games(id, name) VALUES (2, 'minecraft-bedrock');
+
+ALTER TABLE mods ADD COLUMN game_id integer REFERENCES games ON UPDATE CASCADE NOT NULL DEFAULT 1; -- all past ones are minecraft-java
+ALTER TABLE loaders ADD COLUMN game_id integer REFERENCES games ON UPDATE CASCADE NOT NULL DEFAULT 1; -- all past ones are minecraft-java
```
[the remark this answers is not in the document]
Comment (thesuzerain) [reply migrations/20231005230721_dynamic-fields.sql:10]: I didn't see that, but there is some relevance to it here given that we are doing a lot of loader specific stuff (like loader field enumes enums being game-specific but not loader specific). I included the bare minimum implementation for now, I figure in a future PR where we 'fully' implement games this could be done.

File: migrations/20231005230721_dynamic-fields.sql [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+CREATE TABLE loader_field_enums (
+  id serial PRIMARY KEY,
+  game_id integer REFERENCES games ON UPDATE CASCADE NOT NULL DEFAULT 1,
```
[the remark this answers is not in the document]
Comment (thesuzerain) [reply migrations/20231005230721_dynamic-fields.sql:14]: Good point

File: src/routes/v2/admin.rs [hunk 4]
```diff
@@ -319,3 +321,13 @@ pub async fn trolley_webhook(
 
     Ok(HttpResponse::NoContent().finish())
 }
+
+#[post("/_force_reindex")]
```
[the remark this answers is not in the document]
Comment (thesuzerain) [reply src/routes/v2/admin.rs:325]: *theoretically* this should be safe behind an admin wall, but since its only really used for testing anyway, I should probably just put it behind a cfg test wall.

File: src/routes/v2_reroute.rs [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    mut multipart: Multipart,
+    mut headers: HeaderMap,
+    mut closure: impl FnMut(&mut serde_json::Value),
+) -> Result<Multipart, CreateError> {
+    let mut segments: Vec<MultipartSegment> = Vec::new();
+
+    if let Some(field) = multipart.next().await {
+        let mut field = field?;
+        let content_disposition = field.content_disposition().clone(); // This unwrap is okay because we expect every field to have content disposition
```
[the remark this answers is not in the document]
Comment (thesuzerain) [reply src/routes/v2_reroute.rs:79]: Yup thanks

File: src/routes/v2_reroute.rs (hunk 1 above)
Comment (thesuzerain) [reply src/routes/v2_reroute.rs:70]: Yeah definitely.  I think I want to come back to write a lot of tests around this area as I refactor it a little bit. Tbh Im not sure how I feel about the different ways to handle these rerouting functions so far, none of the ideas Ive run through have really felt 100% right

File: src/routes/v2/projects.rs [hunk 6]
```diff
@@ -70,6 +65,73 @@ pub async fn project_search(
...
+    // Loader fields are:
+    // (loader)_(field):(value)
+    // The first _ by convention is used to separate the loader from the field
+    // For each v2 loader, we create a loader field facet for each of these fields that are now loader fields
+    let facets: Option<Vec<Vec<String>>> = if let Some(facets) = info.facets {
+        let facets = serde_json::from_str::<Vec<Vec<&str>>>(&facets)?;
+
+        // "versions:x" => "fabric_game_versions:x", "forge_game_versions:x" ...
+        // They are put in the same array- considered to be 'or'
+        let mut v2_loaders = vec!["fabric", "forge"]; // TODO: populate
```
[the remark this answers is not in the document]
Comment (thesuzerain) [reply src/routes/v2/projects.rs:78]: Yup ty

File: migrations/20231005230721_dynamic-fields.sql [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+INNER JOIN loader_field_enum_values lfev ON m.client_side = lfev.original_id
+WHERE client_side IS NOT NULL AND lfev.enum_id = 1;
+
+INSERT INTO version_fields (version_id, field_id, enum_value) 
+SELECT v.id, 1, m.server_side 
+FROM versions v
+INNER JOIN mods m ON v.mod_id = m.id
+INNER JOIN loader_field_enum_values lfev ON m.client_side = lfev.original_id
+WHERE server_side IS NOT NULL AND lfev.enum_id = 1;
+
```
Comment (Geometrically) [inline migrations/20231005230721_dynamic-fields.sql:85]: drop mods game_versions and loaders

File: src/database/models/project_item.rs [hunk 8]
```diff
@@ -768,31 +754,6 @@ impl Project {
...
-            id as ProjectId,
-            &*crate::models::projects::VersionStatus::iterator().filter(|x| x.is_hidden()).map(|x| x.to_string()).collect::<Vec<String>>()
-        )
-            .execute(&mut *transaction)
-            .await?;
-
-        Ok(())
-    }
-
     pub async fn update_loaders(
```
Comment (Geometrically) [inline src/database/models/project_item.rs:757]: i think you can get rid of this func btw like the above

File: src/models/v2/mod.rs [hunk 9]
```diff
@@ -0,0 +1,2 @@
+// Legacy models from V2, where its useful to keep the struct for rerouting/conversion
+pub mod projects;
```
Comment (Geometrically) [inline src/models/v2/mod.rs:2]: do route restructuring btw

File: src/routes/v2/admin.rs [hunk 10]
```diff
@@ -319,3 +321,13 @@ pub async fn trolley_webhook(
 
     Ok(HttpResponse::NoContent().finish())
 }
+
```
Comment (Geometrically) [inline src/routes/v2/admin.rs:324]: make sure to add `, guard = "admin_key_guard"` here

File: src/routes/updates.rs [hunk 11]
```diff
@@ -95,19 +96,33 @@ pub async fn forge_updates(
     };
 
     for version in versions {
+        // For forge in particular, we will hardcode it to use GameVersions rather than generic loader fields, as this is minecraft-java exclusive
+        // Will have duplicates between game_versions (for non-forge loaders), but that's okay as
+        // before v3 this was stored to the project and not the version
+        let game_version_values: Vec<serde_json::Value> = version
```
Comment (Geometrically) [inline src/routes/updates.rs:102]:
```json 
{ 
   "id": "blah",
    "name": "My version",
    "loaders": ["fabric", "forge"],
    "loader_fields": {
    "client_side": "required",
}
```

restructure loader fields

File: migrations/20231005230721_dynamic-fields.sql [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+INSERT INTO games(id, name) VALUES (2, 'minecraft-bedrock');
+
+ALTER TABLE loaders ADD CONSTRAINT unique_loader_name UNIQUE (loader);
+
+ALTER TABLE mods ADD COLUMN game_id integer REFERENCES games NOT NULL DEFAULT 1; -- all past ones are minecraft-java
+ALTER TABLE loaders ADD COLUMN game_id integer REFERENCES games NOT NULL DEFAULT 1; -- all past ones are minecraft-java
+
+CREATE TABLE loader_field_enums (
+  id serial PRIMARY KEY,
+  game_id integer REFERENCES games NOT NULL DEFAULT 1,
```
Comment (Geometrically) [inline migrations/20231005230721_dynamic-fields.sql:16]: remove this

File: migrations/20231005230721_dynamic-fields.sql [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  enum_name varchar(64) NOT NULL,
+  ordering int NULL,
+  hidable BOOLEAN NOT NULL DEFAULT FALSE
+);
+
```
Comment (Geometrically) [inline migrations/20231005230721_dynamic-fields.sql:21]: make loader fields, loader independent, add separate table to assign version fields to loaders, they can be assigned to multiple loaders

**Review by Geometrically — CHANGES_REQUESTED** [review]

File: src/search/mod.rs [hunk 14]
```diff
@@ -99,6 +97,19 @@ pub struct UploadSearchProject {
...
+       Version fields are stored as:
+           "loader_field": ["value1", "value2", ...]
+       By convention, first underline separates the loader from the field name,
+       and any subsequent underlines may be part of the field name.
+       eg:
+           "fabric_game_versions": ["1.21", "1.22"]
+           "fabric_client_side": ["required"]
+           "fabric_server_side": ["optional"]
+    */
+    pub loader_fields: HashMap<String, Vec<String>>,
```
Comment (Geometrically) [inline src/search/mod.rs:112]: since loader field names are unique I don't think you need to prefix anything here. Also don't see point of flatten

File: src/validate/mod.rs
Comment (Geometrically) [inline src/validate/mod.rs:1]: honestly I think you can just get rid of validators for now and just check for file extensions

**Review by Geometrically — CHANGES_REQUESTED** [review]

File: migrations/20231005230721_dynamic-fields.sql [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+SELECT DISTINCT m.id, c.id 
+FROM mods m
+LEFT JOIN versions v ON m.id = v.mod_id
+LEFT JOIN loaders_versions lv ON v.id = lv.version_id
+LEFT JOIN loaders l ON lv.loader_id = l.id
+CROSS JOIN categories c 
+WHERE m.project_type = (SELECT id FROM project_types WHERE name = 'modpack') AND c.category = l.loader;
+
+-- Non mrpack loaders no longer support modpacks
+DELETE FROM loaders_project_types WHERE joining_loader_id != (SELECT id FROM loaders WHERE loader = 'mrpack') AND joining_project_type_id = (SELECT id FROM project_types WHERE name = 'modpack');
```
Comment (Geometrically) [inline migrations/20231005230721_dynamic-fields.sql:29]: How come the loaders were made into categories? shouldn't they just be per-version dynamic fields?

File: src/routes/v2/images.rs
Comment (Geometrically) [inline src/routes/v2/images.rs:1]: why isn't this an exclusive v3 route? since it was added post v2-prod

File: src/routes/v2/organizations.rs
Comment (Geometrically) [inline src/routes/v2/organizations.rs:1]: same here- this file should be deleted and moved entirely to v3

File: src/routes/v2/teams.rs
Comment (Geometrically) [inline src/routes/v2/teams.rs:1]: orgs related routes in here should also be rerouted

**Review by Geometrically — CHANGES_REQUESTED** [review]

**Review by Geometrically — APPROVED** [review]
